editor/src/core/model/storyboard-utils.ts

```diff
@@ -0,0 +1,129 @@
+import {
+  addFileToProjectContents,
+  getContentsTreeFileFromString,
+  walkContentsTree,
+} from '../../components/assets'
+import { EditorModel } from '../../components/editor/action-types'
+import { EditorState } from '../../components/editor/store/editor-state'
+import { forEachRight, left, right } from '../shared/either'
+import {
+  isUtopiaJSXComponent,
+  UtopiaJSXComponent,
+  utopiaJSXComponent,
+} from '../shared/element-template'
+import { importAlias, isUIJSFile, RevisionsState } from '../shared/project-file-types'
+import { generateUID } from '../shared/uid-utils'
+import { addImport, parseSuccess } from '../workers/common/project-file-utils'
+import { uiJsFile } from './project-file-utils'
+import {
+  BakedInStoryboardUID,
+  BakedInStoryboardVariableName,
+  createSceneFromComponent,
+  createStoryboardElement,
+} from './scene-utils'
+
+export const StoryboardFilePath: string = '/src/storyboard.js'
+
+const PossiblyMainComponentNames: Array<string> = ['App', 'Application', 'Main']
+
+interface ComponentToImport {
+  path: string
+  component: UtopiaJSXComponent
+}
+
+function componentToImport(path: string, component: UtopiaJSXComponent): ComponentToImport {
+  return {
+    path: path,
+    component: component,
+  }
+}
+
+export function addStoryboardFileToProject(editorModel: EditorModel): EditorModel | null {
+  const storyboardFile = getContentsTreeFileFromString(
+    editorModel.projectContents,
+    StoryboardFilePath,
+  )
+  if (storyboardFile == null) {
+    let namedComponentToImport: ComponentToImport | null = null as any
+    let firstComponentToImport: ComponentToImport | null = null as any
+    walkContentsTree(editorModel.projectContents, (fullPath, file) => {
+      // Just in case we ever find a main looking component, skip everything else.
+      if (namedComponentToImport == null) {
+        if (isUIJSFile(file)) {
```

Comment:
I suspect this check is likely to break when we separate storyboard files from regular files in the near future. Right now this check is effectively checking that the file path is `src/app.js`, which _is_ what we want to check for this piece of code, but there's a question around whether we'll still consider that file to be `'UI_JS_FILE'` after we split out storyboard files. Not something to address right now, but something to keep in mind as we might need to come back and change this in the near future.

I'm wondering if what we'll end up with is a type that represents any parsed JSX files, and a type that represents storyboard files :thinking:

Reply:
If the import treats every piece of code as a stock code file that's definitely a big problem as nothing will be parsed for JSX elements. There's not really much in `UIJSFile` that is storyboard specific as we peer inside that representation to see if there is a storyboard.

Reply:
Yeah this is what I mean - right now the project import uses a check to see if the file should be a `'UI_JS_FILE'` based on that same file path (the assumption comes from [this line in `fileTypeFromFileName`](https://github.com/concrete-utopia/utopia/blob/master/editor/src/core/model/project-file-utils.ts#L392)). I can see us ending up with a `JSX_FILE` type or similar which parses all code that includes the JSX pragma, probably in the very near future tbh as we move on to supporting canvas manipulation of arbitrary projects.